### PR TITLE
Attempt Use Of Dates With `idata` Objects 

### PR DESCRIPTION
**Pull request by AFg6K7h4fhy2**
Tests for this will be found here: #29 

This PR ought to be merged before further work on #9 occurs.

### CODE REVIEW

**Comment by AFg6K7h4fhy2** [comment]
Forecasts in this PR are implicitly assumed to be sequential days — not weeks, 2-weeks, a sequence of Saturdays, etc... Other resolution may be supported. Some critical tasks of this PR include (some of these already exist, albeit inadequately):

* Designing a function that adds "dates" as coordinates OR as a group to an `az.InferenceData` object that was generated from AT LEAST `posterior` and `posterior_predictive` groups (should have `observed_data` by default and _might_ not have a `predictions` group). 
* Designing a function that receives an `az.InferenceData` object with date coordinates OR a dates group and converts it to a `polars` dataframe.
* (similar to previous) Designing a function that receives an `az.InferenceData` object with date coordinates OR a dates group and converts it to tidy dataframe.

Key considerations:

* Dates as `coordinates`?
* Dates as groups in `idata` object?
* Names of `idata` groups to expect (e.g. `obs` or `y` in the group `observed_data`)?
* Function in `forecasttools` (not on user-side) for `idata` wo/ dates &rarr; `idata` w/ dates?
* Function in `forecasttools` (not on user-side) for `idata` w/ dates &rarr; [dataframe formats]?

**Comment by AFg6K7h4fhy2** [comment]
Some notes extracted from the experimentation file for adding dates to `idata` objects:

* Function for adding dates to `idata` accept either IS08601 `str` or `datetime` date.
* Any functions that _receives_ a `polars` dataframe with dates ought to have those dates typed `date` (`datetime`).
* `.draw` was artifact from R &rarr; use `draw`.
* There should be name agnostic `idata` reference to dimensions (e.g. reference what is stored in `idata.observed_data["obs_dim_0"]` without using `obs_dim_0`...

NOTE:

```python
# idata.observed_data.sizes["obs_dim_0"] is suggested by Arviz over
# idata.observed_data["obs_dim_0"].size

idata.observed_data.sizes
# output: Frozen({'obs_dim_0': 488})

# idata.observed_data.dims gets one:
# output: FrozenMappingWarningOnValuesAccess({'obs_dim_0': 488})


# ^ "obs_col" name still needed ^
```

* An alternate name, which could be a `forecasttools` default, for a key like `obs_dim_0` could be `day_index`.
* Dates being added to `idata` objects ought to be "as agnostic as possible".
* A function that add to an `idata` object an interval of one-day sequential dates is one possible function for `forecasttools`; other functions could provide other resolutions.
* TPM PR task: 
  * [x] Experimentation using `assign_coords` with dates to `idata`.
  * [x] Refactor function that takes `idata` wo/ dates and converts to `polars` dataframe.
  * [ ] Experimentation using `add_group` with dates to `idata` (shows `plot_ts`).
  * [ ] Using `assign_coords` to create new `.nc` example forecast file.
  * [ ] Refactor function that aggregates values by `epiweek` to use `idata` w/ dates.
  * [ ] Ensure vignette still works.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -16,14 +16,14 @@ repos:
...
-        args: ['--line-length', '79']
+        args: ['--line-length', '50']
         language_version: python3
 -   repo: https://github.com/PyCQA/isort
     rev: 5.13.2
     hooks:
     -   id: isort
         args: ['--profile', 'black',
-               '--line-length', '79']
+               '--line-length', '50']
```
Comment (damonbayer) [inline .pre-commit-config.yaml:26]: Why this change? If anything, I have seen support around CFA for longer line lengths.

**Comment by AFg6K7h4fhy2** [comment]
Questions:

* Should the method that adds dates to `idata` add the dates as type `str`, `datetime`, or `np.datetime`?
* Should the method that takes `idata` w/ dates and converts it to tidy-like dataframes have the output column for dates be of type `str`?

**Comment by AFg6K7h4fhy2** [comment]
Comments on notebook files are appreciated.
Notebooks (save for the FluSight vignette), will not be added to the PR.
I can turn notebooks into vignettes if others desire.

**Comment by AFg6K7h4fhy2** [comment]
Docstrings likely still need to be corrected.

File: forecasttools/daily_to_epiweekly.py [hunk 2]
```diff
@@ -41,7 +44,7 @@ def df_aggregate_to_epiweekly(
         A polars dataframe with draws and dates
         as columns. This dataframe will likely
         have come from an InferenceData object
-        that was passed converted using `forecast_as_df_with_dates`.
+        that was passed converted using `idata_w_dates_to_df`.
```
Comment (damonbayer) [inline forecasttools/daily_to_epiweekly.py:47]: "passed converted" ?

File: forecasttools/recode_locations.py
Comment (damonbayer) [inline forecasttools/recode_locations.py:1]: Seems a bit specific and undocumented to be given its own function.

Comment (AFg6K7h4fhy2) [reply forecasttools/recode_locations.py:1]: There are other, yet unimplemented (partially, as they've not yet been used, e.g. in vignettes) functions that belong in this file. `loc_flusight_code_to_abbr`, `to_location_table_column`, and `location_lookup`. Adding these is out of the scope of this PR. I will make an issue.

File: README.md [hunk 3]
```diff
@@ -184,31 +184,38 @@ Inference data with groups:
...
 
-idata = forecasttools.nhsn_flu_forecast
+# idata with dates as coordinates
+idata_w_dates = forecasttools.nhsn_flu_forecast_w_dates
+
+# idata without dates as coordinates
+idata_wo_dates = forecasttools.nhsn_flu_forecast_wo_dates
 ```
 
 The forecast was generated following the creation of `nhsn_hosp_flu.csv` (see previous section) by running `data.py` with the following added:
```
Comment (damonbayer) [inline README.md:200]: I don't think I am seeing where you generate the idata with dates. Where is that?

Comment (AFg6K7h4fhy2) [reply README.md:200]: These are the cached influenza forecasts that come with forecasttools. 

There is no current vignette that shows the conversion of an `idata` without dates to an `idata` with dates.

There is an experimentation file that contains this conversion. 

I can convert this experimentation file into a vignette if this is desired.

The contents of the experimentation file:

```python
"""
Experimentation file for
examining how idata with
dates operate.
"""

# %% IMPORT LIBRARIES


import xarray as xr

import forecasttools

xr.set_options(
    display_expand_data=False,
    display_expand_attrs=False,
)


# %% GET IDATA WITHOUT DATES

idata_wo_dates = forecasttools.nhsn_flu_forecast_wo_dates

print(idata_wo_dates)

print(idata_wo_dates.observed_data["obs_dim_0"].values[:10])

# %% ADD DATES TO IDATA OBJECT GROUP(S)


start_date_iso = "2022-08-08"
idata_w_dates = forecasttools.add_dates_as_coords_to_idata(
    idata_wo_dates=idata_wo_dates,
    group_dim_dict={
        "observed_data": "obs_dim_0",
        "posterior_predictive": "obs_dim_0",
        "prior_predictive": "obs_dim_0",
    },
    start_date_iso=start_date_iso,
)

print(idata_w_dates)

print(idata_w_dates.observed_data["obs_dim_0"].values[:10])

print(type(idata_w_dates.observed_data["obs_dim_0"].values[0]))


# %% CONVERT DATES IDATA TO DATAFRAME

df_from_idata_w_dates = forecasttools.idata_forecast_w_dates_to_df(
    idata_w_dates=idata_w_dates,
    location="TX",
    postp_val_name="obs",
    postp_dim_name="obs_dim_0",
    timepoint_col_name="date",
    value_col_name="hosp",
)
df_from_idata_w_dates
```

Comment (damonbayer) [reply README.md:200]: You don't need to do a full-blown vignette, but it would be good if all of the objects used in the library were reproducible.

File: forecasttools/idata_w_dates_to_df.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        which to begin the date range for the dimension.
+
+    Returns
+    -------
+    idata
+        The modified InferenceData object with
+        date coordinates assigned to each group.
+    """
+
+    # NOTE: failure mode is if groups don't have the same
```
Comment (damonbayer) [inline forecasttools/idata_w_dates_to_df.py:46]: What does this mean?

File: forecasttools/idata_w_dates_to_df.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+                .select(
+                    pl.date_range(
+                        start=start_date_as_dt,
+                        end=start_date_as_dt
+                        + pl.duration(days=interval_size - 1),
+                        interval="1d",
+                        closed="both",
+                    )
+                )
+                .to_series()
```
Comment (damonbayer) [inline forecasttools/idata_w_dates_to_df.py:72]: Seems like there should be a much more concise way to do this than creating and empty data frame, converting it to a series, and converting that to a numpy array.

File: forecasttools/idata_w_dates_to_df.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        if idata_group is not None:
+            interval_size = idata_group.sizes[dim_name]
+            interval_dates = (
+                pl.DataFrame()
+                .select(
+                    pl.date_range(
+                        start=start_date_as_dt,
+                        end=start_date_as_dt
+                        + pl.duration(days=interval_size - 1),
+                        interval="1d",
```
Comment (damonbayer) [inline forecasttools/idata_w_dates_to_df.py:68]: Shouldn't this be configurable?

File: forecasttools/idata_w_dates_to_df.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    timepoint_col_name: str = "date",
+    value_col_name: str = "hosp",
+) -> pl.DataFrame:
+    """
+    Converts an Arviz InferenceData object
+    into a Polars dataframe contains dates
+    and draws for a given location. The
+    number of rows in the dataframe is equal
+    to the number of posterior predictive
+    samples times the duration of the time
```
Comment (damonbayer) [inline forecasttools/idata_w_dates_to_df.py:103]: What happens to variables that don't change in time?

This is not the typical tidy_draws format, but I'm not sure how the usual format will play with date indexes.

File: README.md (hunk 3 above)
Comment (damonbayer) [reply README.md:200]: I think "reproduce objects in the package" scripts can be minimally-documented notebooks/vignettes.

File: forecasttools/idata_w_dates_to_df.py (hunk 7 above)
Comment (damonbayer) [reply forecasttools/idata_w_dates_to_df.py:103]: It will work fine with date indexes:

``` r
library(tidybayes)
library(tidyverse)
set.seed(5)
n = 10

example <- 
  tibble(
    condition = factor(rep(str_c("A[", seq(today()-4, today(), by = 1), "]"), n)),
    response = rnorm(n * 5, c(0,1,2,1,-1), 0.5)
  ) %>%
  group_by(condition) %>%
  mutate(.chain = 1, .iteration = row_number()) %>% 
  pivot_wider(names_from = condition, values_from = response) %>% 
  mutate(.draw = tidybayes:::draw_from_chain_and_iteration_(.chain, .iteration), .after = .iteration) %>%  tidy_draws()


example %>% spread_draws(A[date])
#> # A tibble: 50 × 5
#> # Groups:   date [5]
#>    date             A .chain .iteration .draw
#>    <chr>        <dbl>  <int>      <int> <int>
#>  1 2024-10-21 -0.420       1          1     1
#>  2 2024-10-21 -0.301       1          2     2
#>  3 2024-10-21  0.614       1          3     3
#>  4 2024-10-21 -0.0695      1          4     4
#>  5 2024-10-21  0.450       1          5     5
#>  6 2024-10-21 -0.147       1          6     6
#>  7 2024-10-21  0.158       1          7     7
#>  8 2024-10-21  0.476       1          8     8
#>  9 2024-10-21  0.775       1          9     9
#> 10 2024-10-21  0.281       1         10    10
#> # ℹ 40 more rows
```

<sup>Created on 2024-10-25 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

File: forecasttools/idata_w_dates_to_df.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+"""
+
+from datetime import datetime
+
+import arviz as az
+import numpy as np
+import polars as pl
+
+
+def add_dates_as_coords_to_idata(
```
Comment (damonbayer) [inline forecasttools/idata_w_dates_to_df.py:16]: Just clarifying - this is intended to be a minimal version of this function, and we will add a more full-featured on soon?

File: forecasttools/idata_w_dates_to_df.py (hunk 4 above)
Comment (damonbayer) [reply forecasttools/idata_w_dates_to_df.py:46]: In general, each variable should be able to have arbitrary dates.

File: forecasttools/idata_w_dates_to_df.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    location: str,
+    postp_val_name: str,
+    postp_dim_name: str,
+    timepoint_col_name: str = "date",
+    value_col_name: str = "hosp",
+) -> pl.DataFrame:
+    """
+    Converts an Arviz InferenceData object
+    into a Polars dataframe contains dates
+    and draws for a given location. The
```
Comment (damonbayer) [inline forecasttools/idata_w_dates_to_df.py:100]: What is the connection between this function and "location?"

File: forecasttools/idata_w_dates_to_df.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+    number of rows in the dataframe is equal
+    to the number of posterior predictive
+    samples times the duration of the time
+    series (end_date - start_date).
+    """
+    # get dates from InferenceData's posterior_predictive group
```
Comment (damonbayer) [inline forecasttools/idata_w_dates_to_df.py:106]: Why does this only operate on the posterior predictive?

**Comment by damonbayer** [comment]
> Comments on notebook files are appreciated. Notebooks (save for the FluSight vignette), will not be added to the PR. I can turn notebooks into vignettes if others desire.

Some more direction here to reviewers would be helpful. Which notebooks should we look at and what kind of feedback would be useful (since you say these will not be part of the PR)?

File: forecasttools/idata_w_dates_to_df.py (hunk 6 above)
Comment (AFg6K7h4fhy2) [reply forecasttools/idata_w_dates_to_df.py:68]: `interval_size` or `interval`? Assuming `interval` since `interval_size` is determined by `idata_group.sizes[dim_name]`. If `interval`, then I agree without a `str` parameter so default is not `1D`.

File: forecasttools/idata_w_dates_to_df.py (hunk 5 above)
Comment (AFg6K7h4fhy2) [reply forecasttools/idata_w_dates_to_df.py:72]: This is what I originally had and thought would work:

```python
interval_size = idata_group.sizes[dim_name]
interval_dates = pl.date_range(
    start=start_date_as_dt,
    end=start_date_as_dt + pl.duration(days=interval_size - 1),
    interval="1d",
    closed="both"
).to_numpy().astype("datetime64[ns]")
```

and this is the error I got

```
AttributeError: 'Expr' object has no attribute 'to_numpy'
```

Perhaps I am using [date_range](https://docs.pola.rs/api/python/stable/reference/expressions/api/polars.date_range.html) incorrectly.

Comment (AFg6K7h4fhy2) [reply forecasttools/idata_w_dates_to_df.py:72]: Silly mistake on my end. Just re-read the documentation. `eager=True` gets me what I want.

```
eager: Evaluate immediately and return a Series. If set to False (default), return an expression instead.
```

**Comment by AFg6K7h4fhy2** [comment]
> Some more direction here to reviewers would be helpful. Which notebooks should we look at and what kind of feedback would be useful (since you say these will not be part of the PR)?

Notebooks:

* `df_from_idata_w_dates_to_tidy_draws`: Shows how to go from InferenceData without dates to a tidy `data.frame` in R with dates and draws.
* `create_idata_flu_forecast_w_dates`: Produces an influenza hospital admissions forecast `idata` object with dates.
* `explore_add_dates_to_idata_via_coords`: Shows how to go from InferenceData without dates to InferenceData with dates, via the `idata` coordinates.
* `explore_add_dates_to_idata_via_groups`: Shows how to go from InferenceData without dates to InferenceData with dates, via the `idata` groups.
* `test_out_idata_w_dates`: Shows how to go from InferenceData without dates to a polars dataframe of draws and dates.

In a sequence of learning, someone would use: 

`explore_add_dates_to_idata_via_coords` &rarr; `test_out_idata_w_dates` &rarr; `df_from_idata_w_dates_to_tidy_draws`. 

If someone didn't know much about NumPyro to Arviz, then `create_idata_flu_forecast_w_dates` would be the place to go.

**Comment by damonbayer** [comment]
@AFg6K7h4fhy2 why don't you think these notebooks should be part of the PR?

I think they could be used as vignettes or reworked as tests.

**Comment by damonbayer** [comment]
@AFg6K7h4fhy2 Perhaps we should sync face to face about this, but I'm confused about the intended goals here. The PR is titled "Attempt Use Of Dates With idata Objects." I suppose that is what the PR does, but I am wondering if we can narrow the focus.

- There is some functionality for "tidying" `InferenceData` objects with dates into a dataframe (`idata_forecast_w_dates_to_df`) but it does not convert them to the `tidy_draws`-like format, as described in https://github.com/CDCgov/forecasttools-py/issues/18. This PR is not linked to that issue, so I maybe it is not intended to address this need. What is the intended use of the format it does convert them to?
- There is also some functionality for adding dates to idata objects (`add_dates_as_coords_to_idata`), but it is not fully-featured.
- There are miscellaneous updates to other functions and their documentation (ex `forecasttools/recode_locations.py`)

**Comment by AFg6K7h4fhy2** [comment]
From face-to-face discussion with Damon:

* PR goal is to create an `idata` object with dates as its coordinates and then verify that this `idata` works with the existing code (primarily the hubverse submission vignette).

**Comment by AFg6K7h4fhy2** [comment]
Not covered in this PR is the task of writing a function that takes an `idata` object and, in a group/coordinate agnostic manner, converts it to a `tidybases::tidy_draws` dataframe, which is a more general utility than is covered by the function `idata_forecast_w_dates_to_df`.

File: forecasttools/idata_w_dates_to_df.py (hunk 6 above)
Comment (AFg6K7h4fhy2) [reply forecasttools/idata_w_dates_to_df.py]: This is more relevant to #34

File: forecasttools/idata_w_dates_to_df.py (hunk 8 above)
Comment (AFg6K7h4fhy2) [reply forecasttools/idata_w_dates_to_df.py:16]: This will be addressed in #34

File: forecasttools/idata_w_dates_to_df.py (hunk 9 above)
Comment (AFg6K7h4fhy2) [reply forecasttools/idata_w_dates_to_df.py:96]: The function `idata_w_dates_to_df` is less general than an eventual function we want for `idata` to `tidy_draws`; right now, `location` as a column is more relevant to the conversation of `idata` eventually to a hubverse and or a ScoringUtils ready dataframe.

File: forecasttools/idata_w_dates_to_df.py (hunk 10 above)
Comment (AFg6K7h4fhy2) [reply forecasttools/idata_w_dates_to_df.py:102]: This uncertainty will be addressed in #18

File: forecasttools/idata_w_dates_to_df.py (hunk 9 above)
Comment (AFg6K7h4fhy2) [reply forecasttools/idata_w_dates_to_df.py:96]: #18 covers the conversion of `idata` to `tidy_draws`.

File: forecasttools/idata_w_dates_to_df.py (hunk 4 above)
Comment (AFg6K7h4fhy2) [reply forecasttools/idata_w_dates_to_df.py:46]: For #34 this matters.

**Review by damonbayer — APPROVED** [review]
LGTM
